File: observalog-java/src/main/java/com/observalog/Degrade.java
package com.observalog;

import java.util.UUID;

// Decision 5: never panic on missing context. Background jobs are legitimate emitters.
// The sys_ prefix signals the brain without requiring special handling in the emitting code.
public final class Degrade {
    private Degrade() {}

    public static void applyGracefulDegradation(LogContext ctx) {
        if (ctx.getTraceId().isEmpty()) {
            // sys_ + 10 hex chars = 14 chars total, fits CHAR(14).
            String uid = UUID.randomUUID().toString().replace("-", "");
            ctx.setTraceId("sys_" + uid.substring(0, 10));
        }
        if (ctx.getUserId().isEmpty()) {
            ctx.setUserId("system");
        }
        if (ctx.getJourneyStage().isEmpty()) {
            ctx.setJourneyStage("system.background.untraced");
        }
    }
}

File: observalog-node/src/degrade.ts
import { v4 as uuidv4 } from 'uuid';
import { LogContextData } from './context';

// Decision 5: never throw on missing context. Background jobs are legitimate.
// The sys_ prefix signals the brain without any special emitter code.
export function applyGracefulDegradation(ctx: LogContextData): void {
    if (!ctx.traceId) {
        const uid = uuidv4().replace(/-/g, '');
        ctx.traceId = 'sys_' + uid.substring(0, 10); // 14 chars total — fits CHAR(14)
    }
    if (!ctx.userId) {
        ctx.userId = 'system';
    }
    if (!ctx.journeyStage) {
        ctx.journeyStage = 'system.background.untraced';
    }
}
